Modules/itertoolsmodule.c

```diff
@@ -3454,12 +3454,81 @@ count_repr(countobject *lz)
                                 lz->long_cnt, lz->long_step);
 }
 
+static PyObject *
+count_peek(countobject *lz, PyObject *Py_UNUSED(ignored))
+{
+    if (lz->cnt != PY_SSIZE_T_MAX) {
+        return PyLong_FromSsize_t(lz->cnt);
+    } else {
+        return lz->long_cnt;
+    }
+}
+
+static PyObject *
+count_consume(countobject *lz, PyObject *seq)
+{
+    PyObject *it, *item;
+    PyObject *(*iternext)(PyObject *);
+
+    it = PyObject_GetIter(seq);
+    if (it == NULL) {
+        return NULL;
+    }
+
+    iternext = *Py_TYPE(it)->tp_iternext;
+    while ((item = iternext(it)) != NULL) {
+        Py_DECREF(item);
+#ifndef Py_GIL_DISABLED
+        if (lz->cnt == PY_SSIZE_T_MAX) {
+            count_nextlong(lz);
+        }
+        else {
+            lz->cnt++;
+        }
+#else
+        // free-threading version
+        // fast mode uses compare-exchange loop
+        // slow mode uses a critical section
+        Py_ssize_t cnt = _Py_atomic_load_ssize_relaxed(&lz->cnt);
+        for (;;) {
```

Comment:
I never bother about the free-threaded version (I don't know how it works, I just assume that I don't have a global lock) but is this "infinite" loop pattern the way to actually wait for the resources to be available? namely,`_Py_atomic_compare_exchange_ssize` would fail because it's not ready, or something like that?

Reply:
Yes, under the hood it uses something similar to this: https://learn.microsoft.com/en-us/cpp/intrinsics/interlockedcompareexchangepointer-intrinsic-functions?view=msvc-170

Reply:
In simple words, it is an atomic operation that checks 2 values in memory and if they are equal, then sets one of them to a new value. Otherwise it does nothing and returns fail.

Reply:
Actually, I wasn't asking about the atomic functions themselves (that I know about); what I was wondering was about the `for (;;)` instead of a possibly other way to acquire a lock / wait. But thanks anyway!